### DISCUSSION TITLE
How to use with typescript

### DISCUSSION THREAD

**Discussion by santigarcor** [mercadopago/sdk-js#204] [Q&A] [answered]
I installed the sdk on my app and thew types defintions, i'm using TS but I'm not sure how to get the types working I keep getting this error:

`Property 'MercadoPago' does not exist on type 'Window & typeof globalThis'.ts(2339)`

<img width="722" alt="image" src="https://github.com/mercadopago/sdk-js/assets/7565417/93edc006-fca2-4d56-9af1-e6ae9488063f">

Also, I'm not able to import the type definitions to be able to do something. Can anyone help me with this?

**Comment by francineemilia** [comment]
Hi @santigarcor,

Could you please send the more of the code - with the imports and/or script tags - without private data (for example: without your public key)?

**Reply by santigarcor** [reply]
I'm using Vue so here's the code:

```ts
  import { loadMercadoPago } from "@mercadopago/sdk-js"

  const config = useRuntimeConfig()

  const mp = ref<any>(null)  // Here I would like to be able to set the type to the MercadoPago instance
  const bricksBuilder = ref<any>(null)  // Here I would like to be able to set the type to the brick builder
  const cardPaymentBrickController = ref<any>(null) // Here I would like to be able to set the type to the controller

  onMounted(async () => {
    await loadMercadoPago()
    mp.value = new window.MercadoPago(config.mercadoPagoPublicKey) // Here is where i'm getting the undefined error
    bricksBuilder.value = mp.value.bricks()
  })
```

My package json looks like this:
```json
{
  "name": "nuxt-app",
  "private": true,
  "type": "module",
  "scripts": {
    "build": "nuxt build",
    "dev": "nuxt dev",
    "generate": "nuxt generate",
    "preview": "nuxt preview",
    "postinstall": "nuxt prepare"
  },
  "devDependencies": {
    "@nuxt/devtools": "^1.0.8",
    "@nuxtjs/i18n": "^8.2.0",
    "@nuxtjs/tailwindcss": "^6.11.4",
    "@tailwindcss/aspect-ratio": "^0.4.2",
    "@tailwindcss/forms": "^0.5.7",
    "@tailwindcss/typography": "^0.5.10",
    "@types/node": "^20.11.24",
    "@types/uuid": "^9.0.8",
    "@types/vue-tel-input": "^2.1.6",
    "nuxt": "^3.11.2",
    "postcss-import": "^16.0.1",
    "prettier": "^3.2.5",
    "prettier-plugin-tailwindcss": "^0.5.11",
    "@types/mercadopago-sdk-js": "^2.0.15",
    "typescript": "^5.4.2"
  },
  "dependencies": {
    "@headlessui/vue": "^1.7.19",
    "@heroicons/vue": "^2.1.1",
    "@mercadopago/sdk-js": "^0.0.3",
    "@pinia/nuxt": "^0.5.1",
    "date-fns": "^3.6.0",
    "pinia": "^2.1.7",
    "simplebar": "^6.2.5",
    "uuid": "^9.0.1",
    "vue-currency-input": "^3.1.0",
    "vue-tel-input": "^8.3.1"
  },
  "overrides": {
    "vue": "latest"
  }
}

```

Let me know ifg you need something else

**Reply by francineemilia** [reply]
Hi @santigarcor,

Thanks for send more from your code. In fact, your error occur because you need to type window at your project. There is some ways to do that. Our recommendation is type window like this
```ts 
declare global {
    interface Window {
        MercadoPago: any;
    }
}
````
And use our [type lib](https://www.npmjs.com/package/@types/mercadopago) to help you typing the functions that you will use from SDK.

You can use MercadoPago as `window['MercadoPago'] = ... ` but you'll need to check if the value exists before using, with some code like `if(window["MercadoPago"]){...}`

**Reply by santigarcor** [reply]
@francineemilia I already have installed that type lib, but then how I make it work with the code, because I'm not able to import anything from that lib.

I think adding a guideline for TS on the docs or repo readme will help others in the future.

**Reply by santigarcor** [reply]
@francineemilia Also, It should be `MercadoPago: any;` it should be the actual type defined for mercado pago, that's what i'm talking about it will get rid of the error message but still is a "useless" solution because I'm not getting the type hinting/checking.

**Reply by santigarcor** [reply]
@francineemilia I'm checking your type library and it doesn't export anything. On the index.d.ts file it declares the namespaces but then they are not exported, so how should I use them?

**Reply by francineemilia** [reply]
Hi @santigarcor
You can access our type library through namespace using `mercadopagocore` to access all types from the library.

As you can see at this video, installing de SDK and the type library - as you already did - should solve this questions. The video is a little bit different from your case cause it isn't using Vue and doesn't need typing the `const mp` before define it.

https://github.com/mercadopago/sdk-js/assets/11588420/ce25e9fa-29ba-41ca-a56e-6f16a99ce337

**Reply by santigarcor** [reply]
Thanks for the video but that doesn't work on a Nuxt project. Seems like something is missing because other types from other packages work fine.

**Reply by francineemilia** [reply]
@santigarcor I started a nuxt project and, as far as I could go, there is a tsconfig that exclude most archives from node_modules, where our type library is installed. I suggest add at the include clause the line `"node_modules/@types/mercadopago-sdk-js"`, but the root cause is that exclude clause from nuxt.

[Edit] Fixing my coment, add `"@types/mercadopago-sdk-js"` to the types array at the compilerOptions from tsconfig

**Reply by santigarcor** [reply]
But that's weird because I installed other @types packages for example for lodash and uuid and they work fine, I didn't have to do anything special

**Reply by AleFossati** [reply]
Hello @santigarcor, I have also tried to use `@types/mercadopago-sdk-js` within a nuxt project and I've noticed that `.ts` files inside the `server` folder, have autocomplete and intellisense perfectly fine, just as shown in the previous video.

However, `.ts` files created at other folders, don't get the same behaviour. Since I haven't used nuxt before, I'm not an expert to provide guidance on how it should be done. But I believe this indicates that it is a project typescript local configuration problem, rather than an issue with `@types/mercadopago-sdk-js`.

If you need the mercadopago namespaces to be available in other folders besides "server", maybe the tsconfig `include` keyword will be your way to go.

**Reply by santigarcor** [reply]
But as I said before, I installed other `@types/` packages for `lodash` and `uuid` and they work fine out of the box. So that statement is not 100% true IMO. Also, I've compared both server and default nuxt `tsconfig` file and the include section is almost the same.

**Reply by AleFossati** [reply]
The other `@types/` packages that you mentioned, they work in every file across your project, or you use them only inside the server folder? This is still not clear to me 🤔

**Reply by santigarcor** [reply]
I'm not using the server section at all. I use them on my `.vue` and `.ts` files that are not part of the server folder.

**Reply by AleFossati** [reply, marked as answer]
I've installed `@types/lodash` here as well and it does have the same behaviour that `@types/mercadopago-sdk-js` has in Nuxt, unfortunately this is an inconvenient **caused by the Nuxt configuration**. See the video below.

https://github.com/mercadopago/sdk-js/assets/104934463/aed98ccc-9546-44a7-bdaa-ebb39669c5d1

The `_` is available only inside the `server` folder and yes, you can use it in other folders, but you have to explicitly import it like `import _ from 'lodash'`, see the video below.

https://github.com/mercadopago/sdk-js/assets/104934463/cb58a466-a6fe-44f3-8ae9-012ddab3c0d5

The problem here is that lodash functions can be imported, so it does make sense that it works this way.

In the other hand, the [@mercadopago/sdk-js npm package](https://www.npmjs.com/package/@mercadopago/sdk-js) is only a wrapper of the actual mercadopago/sdk-js, which is available via cdn at https://sdk.mercadopago.com/js/v2. MercadoPago functions, bricks, etc, cannot be imported, they reside inside the `window` object only.

What you can do to get autocomplete and intellisense from your IDE, is explicitly include `@types/mercadopago-sdk-js` in your `.tsconfig` so it'll be available across all your project, or you can add the traditional [triple slash](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html) in the file where you want to use it.
```
/// <reference path="node_modules/@types/mercadopago-sdk-js/index.d.ts" />
```

https://github.com/mercadopago/sdk-js/assets/104934463/f7682769-3846-4923-82f9-a9724ae5dd43

**Reply by santigarcor** [reply]
Thanks for that explanation, now it works fine. The issue, as you said is that mercadopago doesn't export something to import that's why you need to do what you just mentioned.

Sorry for the inconvenience and thanks for the help.
